### DISCUSSION TITLE
Initial set of tracks

### DISCUSSION THREAD

**Discussion by bautistaaa** [typehero/typehero#420] [Announcements]
## Overview
Here are the initial set of tracks that were discussed that we would consider must haves for launch. Titles are not set in stone but the overall theme still stands for each one. Before seeking additional help the current expectation is to create a couple high quality challenges for each track that can be used as a reference for future contributors.

### Tracks 
- Build Built In Helpers (Pick etc)
- Some Expert Track.. Title TBD
- JS Built In Methods
- Understanding Typescript Syntax
- Typescript Foundations
  1. unions
  2. extends
  3. infer
  4. type annotations
  5. array
  6. readonly
  7. map types
  8. etc
  
### How to submit.
My initial thought is shared TS Playgrounds in the thread or on discord but open to suggestions

### Additional notes
- a track could be incremental progress towards understanding some topic
  - would need a way to hide those tracks so they couldnt be taken out of context
  - will probably want to avoid for now and leverage the foundation we have currently but could go down this route

**Comment by bautistaaa** [comment]
here is how to use our type helper thanks to @ggrandi !

https://tsplay.dev/mpQMMW

**Comment by bautistaaa** [comment]
set up an issue template to submit challenges: https://github.com/typehero/typehero/issues/new?assignees=bautistaaa&labels=challenge&projects=&template=new-challenge.yml&title=%5BChallenge%5D%3A+

**Comment by bautistaaa** [comment]
@MichiganTypeScript any thoughts on the initial set of challenges that we can start building out/planning

**Reply by bautistaaa** [reply]
i didnt know your other github handle >.<

**Reply by dimitropoulos** [reply]
hi! :)

**Reply by dimitropoulos** [reply]
looks good: the builtins are in a few groups.  Here's how I'd split it up:

### Group 1: Easy
- [Partial](https://github.com/microsoft/TypeScript/blob/b12af0fa2bbd4b015e59adcfb49988cea7f919a1/src/lib/es5.d.ts#L1551)
- [Required](https://github.com/microsoft/TypeScript/blob/b12af0fa2bbd4b015e59adcfb49988cea7f919a1/src/lib/es5.d.ts#L1558)
- [Readonly](https://github.com/microsoft/TypeScript/blob/b12af0fa2bbd4b015e59adcfb49988cea7f919a1/src/lib/es5.d.ts#L1565)
- [Pick](https://github.com/microsoft/TypeScript/blob/b12af0fa2bbd4b015e59adcfb49988cea7f919a1/src/lib/es5.d.ts#L1572)
- [Record](https://github.com/microsoft/TypeScript/blob/b12af0fa2bbd4b015e59adcfb49988cea7f919a1/src/lib/es5.d.ts#L1579C40-L1579C40)

### Group 2: Medium
- [Exclude](https://github.com/microsoft/TypeScript/blob/b12af0fa2bbd4b015e59adcfb49988cea7f919a1/src/lib/es5.d.ts#L1586)
- [Extract](https://github.com/microsoft/TypeScript/blob/b12af0fa2bbd4b015e59adcfb49988cea7f919a1/src/lib/es5.d.ts#L1591)
- [Omit](https://github.com/microsoft/TypeScript/blob/b12af0fa2bbd4b015e59adcfb49988cea7f919a1/src/lib/es5.d.ts#L1596)
- [Parameters](https://github.com/microsoft/TypeScript/blob/b12af0fa2bbd4b015e59adcfb49988cea7f919a1/src/lib/es5.d.ts#L1606)
- [ReturnType](https://github.com/microsoft/TypeScript/blob/b12af0fa2bbd4b015e59adcfb49988cea7f919a1/src/lib/es5.d.ts#L1616)

### Group 3: Hard
- [ConstructParameters](https://github.com/microsoft/TypeScript/blob/b12af0fa2bbd4b015e59adcfb49988cea7f919a1/src/lib/es5.d.ts#L1611C6-L1611C27)
- [InstanceType](https://github.com/microsoft/TypeScript/blob/b12af0fa2bbd4b015e59adcfb49988cea7f919a1/src/lib/es5.d.ts#L1621)
- [Awaited](https://github.com/microsoft/TypeScript/blob/b12af0fa2bbd4b015e59adcfb49988cea7f919a1/src/lib/es5.d.ts#L1536C6-L1536C13)
- [ThisParameterType](https://github.com/microsoft/TypeScript/blob/b12af0fa2bbd4b015e59adcfb49988cea7f919a1/src/lib/es5.d.ts#L306C6-L306C23)
- [OmitThisParameter](https://github.com/microsoft/TypeScript/blob/b12af0fa2bbd4b015e59adcfb49988cea7f919a1/src/lib/es5.d.ts#L311C6-L311C23)
- [ArrayLike](https://github.com/microsoft/TypeScript/blob/b12af0fa2bbd4b015e59adcfb49988cea7f919a1/src/lib/es5.d.ts#L1543C11-L1543C20)
- [Promise](https://github.com/microsoft/TypeScript/blob/b12af0fa2bbd4b015e59adcfb49988cea7f919a1/src/lib/es5.d.ts#L1516C14-L1516C14)
- [PromiseLike](https://github.com/microsoft/TypeScript/blob/b12af0fa2bbd4b015e59adcfb49988cea7f919a1/src/lib/es5.d.ts#L1503C11-L1503C22)

#### Group 3.a: intrinsics
These are hardcoded into the compiler with the `intrinsic` keyword, but it's a good exercise to try to implement them manually.
- [Uppercase](https://github.com/microsoft/TypeScript/blob/b12af0fa2bbd4b015e59adcfb49988cea7f919a1/src/lib/es5.d.ts#L1626C6-L1626C15)
- [Lowercase](https://github.com/microsoft/TypeScript/blob/b12af0fa2bbd4b015e59adcfb49988cea7f919a1/src/lib/es5.d.ts#L1631)
- [Capitalize](https://github.com/microsoft/TypeScript/blob/b12af0fa2bbd4b015e59adcfb49988cea7f919a1/src/lib/es5.d.ts#L1636)
- [Uncapitalize](https://github.com/microsoft/TypeScript/blob/b12af0fa2bbd4b015e59adcfb49988cea7f919a1/src/lib/es5.d.ts#L1641)

### Group 4: Extreme
- [NonNullable](https://github.com/microsoft/TypeScript/blob/b12af0fa2bbd4b015e59adcfb49988cea7f919a1/src/lib/es5.d.ts#L1601)
- [ThisType](https://github.com/microsoft/TypeScript/blob/b12af0fa2bbd4b015e59adcfb49988cea7f919a1/src/lib/es5.d.ts#L1646C11-L1646C19)

**Reply by bautistaaa** [reply]
thanks for the thorough reply!

I don't need a detailed breakdown of the others but curious on what you'd like to see on these 

```
JS Built In Methods
Understanding Typescript Syntax
Typescript Foundations
```

**Reply by dimitropoulos** [reply]
I'd like to see these:

### JS Built-In

Now, some of these are extremely hard but as far as I know all of them are possible to type to a pretty high degree.  All are worth understanding, I'd think, at some point along the way.

- String: charAt, concat, indexOf, substring, toLowerCase, toUpperCase, split, trim, replace, match
- Array: push, pop, shift, unshift, concat, join, slice, splice, forEach, map
- Object: keys, values, entries, assign
- Numbers: toFixed, toPrecision, parseInt, parseFloat
  
### Understanding TypeScript Syntax

Here's a good place to start.  Every single syntax to write a function.  One. By. One.  This is something we who have been around the block take for granted but is a big stopping point for many newcomers.  You can declare functions in typescript interfaces, in objects themselves, with object shorthand notation, without object shorthand notation, in classes as methods, in classes as lambdas, type aliases, arrow functions, you can type a const with a type alias, etc. etc. etc.

Try to sit down and think of literally every possible syntax to write this function (and/or the type for this function): `const addOne = x => x + 1` and I think you'll be shocked how many there are.

### TypeScript foundations

This one.. honestly... can just be an interactive reworking of [the handbook](https://www.typescriptlang.org/docs/handbook/2/basic-types.html).  Orta (and many more) did a great job of putting it together from a "for people looking for the basics" standpoint.  Why fix what's not broken?

**Reply by bautistaaa** [reply]
i like the interactive handbook idea! will need to make that a 2.0 but i think with all the other tracks + leveraging type challenges for some its puts us in a good spot. just need to make some good examples of these for others to use as "examples"

i added a new issue flow for track submissions if you had any feedback there

**Comment by QuiiBz** [comment]
Adding more ideas for tracks, probably not needed for the MVP but would be great to have in the future:

### Writing TS in the real world

Quite important topic that focuses on writing application code in a typesafe way. Things like:
- Typesafe `switch`
- Why prefer `const` objects instead of enums in most cases
- ...

### How X library works

Recreating popular libraries (in an easy way) that have interesting TS patterns to understand how they work. Things like:
- tRPC
- Zod
- ...
